File: ProductofArrayExceptSelf.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int left = 1;
        int right = 1;
        int result[] = new int[n];
        for(int i = 0; i < nums.length; i++){
            result[i] = 1;
        }
        for(int i = 0; i < nums.length; i++){
            result[i] *= left;
            left *= nums[i]; 
        }
        for(int i = nums.length-1; i >= 0; i--){
            result[i] *= right;
            right *= nums[i];
        }
        return result;
    }
}

File: ProductofArrayExceptSelf.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        numlen = len(nums)
        left = 1
        right = 1
        vals = [1] * numlen
        for i in range(1, numlen):
            left *= nums[i-1]
            vals[i] *= left
        for i in range(numlen-2, -1, -1):
            right *= nums[i+1]
            vals[i] *= right
        return vals

        